README.md
# R11

[![HitCount](http://hits.dwyl.io/PhDMeiwp/R11.svg)](http://hits.dwyl.io/PhDMeiwp/R11)

## Installing "R11" package in R

Get the development version from github:

    install.packages("devtools")
	devtools::install_github("PhDMeiwp/R11@master", force = TRUE)
	library(R11)

## Usage
	data_summary(x, digits = 3)
	
	ggbar_MeanSD(x, y, 
			fill.bar = NULL, width.bar = 0.5, 
			col.border = NULL, size.border = 1, 
			alpha = 1, 
			width.errorbar = 0.3, col.errorbar = NULL, size.errorbar = 1)

## Examples
	# library(R11)
    x<-c(1,3,4,6,8,11,13,15,40,50,100) 
    data_summary(x, digits = 3)
	
*then output as follows:*

 <img src="images/data_summary.jpg" width="490">
	
	# library(R11)
	y = iris$Sepal.Length
	x = iris$Species
	ggbar_MeanSD(x, y)
	
*then output as follows:*
 
  <img src="images/ggbar_MeanSD.png" width="490">

NEWS
Changes in version 0.0.91
------------------------------------------------------------------------------
Date: 2018-09-15
* add `ggbar_MeanSD()` function
* add `theme_ggR11()` function



R/ggbar.R
#' Bar Charts Showing Mean and SD
#'
#' @description Dataset contains only 'y' and 'factor(x)'. The heights of the bars represent y-values against x-factors.
#' @import tidyverse
#' @import xfun
#' @import ggplot2
#'
#' @export
#' @examples
#' library(R11)
#' y = iris$Sepal.Length
#' x = iris$Species
#'
#' # vertical
#' ggbar_MeanSD(x, y)
#'
#' # horizontal
#' ggbar_MeanSD(x, y, alpha = 0.5) +
#'   coord_flip()
#'
#' # set colors
#' ggbar_MeanSD(x, y, fill.bar = c("grey10", "grey50", "grey70"))+
#'   theme_bw()
#'
#' # set ylim using coor_cartesian()
#' ggbar_MeanSD(x, y, alpha = 0.5) +
#'    coord_cartesian(ylim = c(0, 10))
#'
#' @seealso \code{\link[ggplot2]{geom_bar}}, \code{\link[tidyverse]}

ggbar_MeanSD <- function(x, y,
                    fill.bar = NULL,
                    width.bar = 0.5,
                    col.border = NULL,
                    size.border = 1,
                    alpha = 1,
                    width.errorbar = 0.3,
                    col.errorbar = NULL,
                    size.errorbar = 1)
  {

  if (requireNamespace("xfun", quietly = TRUE)){
    xfun::pkg_attach2(c("ggplot2", "tidyverse"))
  }


  data <- data.frame(x,y)
  data2 <- data %>%
    group_by(x) %>%
    summarise(y.mean = mean(y), sd = sd(y))

# color.default
  color.default = c("gray0", "red", "green", "blue", "cyan",
                    "magenta", "yellow", "gray", "purple", "darkorange",
                    "gold", "pink", "gray50", "plum", "red4",
                    "green4", "blue4", "cyan4", "magenta4", "cornsilk")
  color.default = color.default[1:length(levels(x))]
  if (is.null(fill.bar)){fill.bar = color.default}else{fill.bar = fill.bar}
  if (is.null(col.border)){col.border = fill.bar}else{col.border = col.border}
  if (is.null(col.errorbar)){col.errorbar = fill.bar}else{col.errorbar = col.errorbar}

# plot
  ggplot(data2) +
    geom_bar( aes(x=x, y=y.mean), stat="identity",
              fill=fill.bar, width = width.bar, colour=col.border, size=size.border,alpha=alpha)+
    geom_errorbar( aes(x=x, ymin=y.mean-sd, ymax=y.mean + sd),
                   width=width.errorbar, colour=col.errorbar, alpha=alpha, size=size.errorbar)+
    labs(x=deparse(substitute(x)), y = deparse(substitute(y)))

  }


R/theme_ggR11.R
#'
#' @title Default theme used in all `R11` package ggplots
#' @name theme_ggR11
#' @author Indrajeet Patil, Weiping Mei
#' @description Common theme used across all ggplots generated in `R11`
#'   and *assumed* by the author to be aesthetically pleasing to the
#'   user/reader.
#'
#' @param ggtheme A function, `ggplot2` theme name. Default value is
#'   `ggplot2::theme_bw()`. Any of the `ggplot2` themes, or themes from
#'   extension packages are allowed (e.g., `ggthemes::theme_economist()`,
#'   `hrbrthemes::theme_ipsum_ps()`, `ggthemes::theme_fivethirtyeight()`).
#' @param ggR11.layer Logical that decides whether `theme_ggR11`
#'   theme elements are to be displayed along with the selected `ggtheme`
#'   (Default: `TRUE`).
#'
#' @return A `ggplot2` object with the `theme_ggR11` theme.
#'
#' @import ggplot2
#' @importFrom grid unit
#'
#' @references
#' \url{https://indrajeetpatil.github.io/ggR11/articles/theme_ggR11.html}
#'
#' @export
#'

theme_ggR11 <- function(ggtheme = ggplot2::theme_bw(), ggR11.layer = TRUE) {
  if (isTRUE(ggR11.layer)) {
    ggtheme +
      ggplot2::theme(
        axis.title.x = ggplot2::element_text(size = 12, face = "bold"),
        strip.text.x = ggplot2::element_text(size = 12, face = "bold"),
        strip.text.y = ggplot2::element_text(size = 12, face = "bold"),
        strip.text = ggplot2::element_text(size = 12, face = "bold"),
        axis.title.y = ggplot2::element_text(size = 12, face = "bold"),
        axis.text.x = ggplot2::element_text(size = 12, face = "bold"),
        axis.text.y = ggplot2::element_text(size = 12, face = "bold"),
        axis.line = ggplot2::element_line(),
        legend.text = ggplot2::element_text(size = 12),
        legend.title = ggplot2::element_text(size = 12, face = "bold"),
        legend.title.align = 0.5,
        legend.text.align = 0.5,
        legend.key.height = grid::unit(x = 1, units = "line"),
        legend.key.width = grid::unit(x = 1, units = "line"),
        plot.margin = grid::unit(x = c(1, 1, 1, 1), units = "lines"),
        panel.border = ggplot2::element_rect(
          color = "black",
          fill = NA,
          size = 1
        ),
        plot.title = ggplot2::element_text(
          color = "black",
          size = 13,
          face = "bold",
          hjust = 0.5
        ),
        plot.subtitle = ggplot2::element_text(
          color = "black",
          size = 11,
          face = "bold",
          hjust = 0.5
        )
      )
  } else {
    ggtheme
  }
}



R/eg_ggsave300dpi.R
#' An Example for 'ggplot' with Errorbar
#'
#' @description An exmaple.
#' @export
#' @examples
#' library(R11)
#' library(ggplot2)
#' x=c(1,3,8)
#' y=c(5,6,1)
#' dat=data.frame(x,y)
#'
#' ggplot(dat, aes(x = x, y = y)) + geom_line() + theme_bw()
#' p <- last_plot()
#' ggsave("plot1.pdf", p, width = 3.15, height = 3.15)
#' ggsave("plot2.tiff", p, width = 4, height = 3, units="cm", dpi=300)
#' dev.off()

#' # ggsave(filename, plot = last_plot(), device = NULL, path = NULL,
#' #       scale = 1, width = NA, height = NA, units = c("in", "cm", "mm"),
#' #       dpi = 300, limitsize = TRUE, ...)
#' ### 'device' can be c("eps", "ps", "tex", "pdf", "jpeg", "tiff", "png", "bmp", "svg", "wmf").
#' ###  'units' (of 'width', 'height') can be c("in", "cm", or "mm").
#' ###  'scale' refer to "multiplicative scaling factor".
#' @seealso \code{\link[ggplot2]}

eg_ggsave<-function(){?eg_ggsave}


R/ggplot_errorbar.R
#' An Example for 'ggplot' with Errorbar
#'
#' @description An exmaple.
#' @export
#' @examples
#' library(R11)
#' # ggplot_errorbar()
#' meanx <- c(-20,-18,-15)
#' meany <- c(6,8,9)
#' sdx <- c(0.7, 1, 0.9)
#' sdy <- c(1.2, 1.5, 0.8)
#' Groups <- c('group1', 'group2', 'group3')
#' dat<-data.frame(meanx, meany, sdx, sdy, Groups)
#' View(dat)
#'
#' library(ggplot2)
#' cols<-c("grey10", "blue", "red")
#'
#' ##--------------------------------------------
#' #      Case1: geom_barplot and geom_errorbar
#' ##--------------------------------------------
#' p<- ggplot(data=dat,mapping = aes(x=dat$Groups, y=dat$meany, fill=factor(dat$Groups)))+
#'        geom_col(width=0.3)+
#'        scale_fill_manual(values = cols)+
#'        scale_y_continuous(limits = c(0,13), expand = c(0,0))+  # expand = c(0,0) means no space between ymin and x-axis.
#'        xlab('Groups')+ylab('x value')+ theme_bw()
#' p <- p + theme(panel.grid = element_blank())
#' p
#'
#' q <- p + geom_errorbar(aes(ymin=dat$meany-dat$sdy, ymax=dat$meany+dat$sdy), width=0.15)
#' q
#'
#' q + annotate(geom="text", x=dat$Groups[1], y=dat$meany[1]+dat$sdy[1]+0.5, label="b", color="black") +
#'     annotate(geom="text", x=dat$Groups[2], y=dat$meany[2]+dat$sdy[2]+0.5, label="a", color="black") +
#'     annotate(geom="text", x=dat$Groups[3], y=dat$meany[3]+dat$sdy[3]+0.5, label="a", color="black")
#'
#'
#' ##--------------------------------------------
#' #     Case2: geom_point and double errorbars
#' ##--------------------------------------------
#' p<- ggplot(data=dat,mapping = aes(x=dat$meanx, y=dat$meany))+
#'        geom_point(aes(colour=factor(dat$Groups)))+
#'        scale_color_manual(values = cols)+
#'        scale_y_continuous(limits = c(0,13), expand = c(0,0))+
#'        xlab('Groups')+ylab('x value')+ theme_bw()
#' p
#' q <- p +
#'     geom_errorbar(aes(ymin=dat$meany-dat$sdy, ymax=dat$meany+dat$sdy), width=0.3)+
#'     geom_errorbarh(aes(xmin=dat$meanx-dat$sdx, xmax=dat$meanx+dat$sdx), height=0.3)
#' q
#'
#' q + annotate(geom="text", x=dat$meanx[1], y=dat$meany[1]+dat$sdy[1]+0.5, label="b", color="black") +
#'     annotate(geom="text", x=dat$meanx[2], y=dat$meany[2]+dat$sdy[2]+0.5, label="a", color="black") +
#'     annotate(geom="text", x=dat$meanx[3], y=dat$meany[3]+dat$sdy[3]+0.5, label="a", color="black")
#'
#' # More links
#' # https://ggplot2.tidyverse.org/reference/geom_point.html
#' # https://ggplot2.tidyverse.org/reference/geom_linerange.html
#' # https://groups.google.com/forum/#!topic/ggplot2/aiJKyg8U7QM
#' @seealso \code{\link[ggplot2]}

ggplot_errorbar<-function(){?ggplot_errorbar}


R/barplot_or_boxplot.R
#' An Example for Choosing 'barplot' or 'boxplot'
#' 
#' @description An exmaple.
#' @export
#' @examples 
#' library(R11)
#' # barplot_or_boxplot()
#' a<-1:5
#' b<-2:14
#' c<-c(1,5,6,15,25,14,20)
#' a<-as.numeric(a)
#' b<-as.numeric(b)
#' x<-c(a,b,c)
#' g <- c('group1', 'group2', 'group3')
#' group<-rep(g, c(5,13,7))
#' dat<-data.frame(x,group)
#' View(dat)
#' 
#' tapply(dat$x, dat$group, shapiro.test) # Normal distribution test
#' 
#' ##-----------------------------------------
#' #     Test of Homogeneity of Variances
#' ##-----------------------------------------
#' 
#' bartlett.test(dat$x~dat$group) 
#' 
#' # Levene's Test (center = median)
#' library(car)
#' leveneTest(dat$x, dat$group) # P-value>0, then boxplot
#' 
#' boxplot(dat$x~dat$group, ylim=c(0,30))
#' 
#' ##-------------------------------------------------
#' #     Significance test and its post-hoc test
#' ##-------------------------------------------------
#' 
#' kruskal.test(dat$x~dat$group)
#' 
#' library(dunn.test)
#' # Dunn's test of multiple comparisons using rank sums
#' dunn.test(x=dat$x, g = dat$group,list = T)
#' 
#' tapply(dat$x, dat$group, median)
#' 
#' # label significance followed dunn.test results 
#' 
#' text(1, max(dat$x[dat$group=="group1"])+1,"b")
#' text(2, max(dat$x[dat$group=="group2"])+1,"a")
#' text(3, max(dat$x[dat$group=="group3"])+1,"a")
#' 
#' ##-------------------------------------------------
#' #        Alternatively, use ggplot2
#' ##------------------------------------------------- 
#' 
#' library(ggplot2)
#' cols<-c("grey50", "blue", "red")
#' 
#' p<- ggplot(data=dat,mapping = aes(x=dat$group, y=dat$x))+
#'        geom_boxplot(fill=cols, alpha=0.7)+
#'        ylim(0,30)+
#'        xlab('Groups')+ylab('x value')+ theme_bw()
#' p
#' 
#' p + geom_jitter()
#' 
#' p + annotate(geom="text", x=1, y=max(dat$x[dat$group=="group1"])+1, label="b", color="black") + 
#'     annotate(geom="text", x=2, y=max(dat$x[dat$group=="group2"])+1, label="a", color="black") + 
#'     annotate(geom="text", x=3, y=max(dat$x[dat$group=="group3"])+1, label="a", color="black")
#'     
#' # More links
#' # http://www.sthda.com/english/wiki/ggplot2-texts-add-text-annotations-to-a-graph-in-r-software
#' # http://t-redactyl.io/blog/2016/04/creating-plots-in-r-using-ggplot2-part-10-boxplots.html
#' @seealso \code{\link[stats]{bartlett.test}}, \code{\link[car]{leveneTest}}, \code{\link[DescTools]{LeveneTest}}, \code{\link[dunn.test]{dunn.test}} 

barplot_or_boxplot<-function(){?barplot_or_boxplot}


R/data_summary.R
#' Data Summary
#' 
#' @description The results include sumple number (N), mean, varicance (Var), standard deviation (sd), standard error (se, = sd/sqrt(n)), coefficient of variation (CV, %),
#'  median, median absolute deviation (mad), the first quartile (Q1), the third quartile (Q3), interquartile range (IQR, = Q3-Q1), minimum, maximum, range (Max-Min), corrected sum or squares (CSS), uncorrected sum or squares (USS).
#' @import stats
#' @export
#' @examples 
#' library(R11)
#' x<-c(1,3,4,6,8,11,13,15,40,50,100) 
#' data_summary(x)
#' 
#' # Not run------------------
#' shapiro.test(x)  
#' # if p-value>0.05 (i.e. normal distribution), use Mean +/- sd; 
#' # else, use Meadian [Q1-Q3] or Median [IQR].
#' 
#' # Alternatively, mean (95% CI for the mean) and median (95%CI for the median) are also used in some scientific papers.
#' # In general, 95%CI is not recommended, unless required by the target journal.
#' library(DescTools)
#' DescTools::MeanCI(x)
#' DescTools::MedianCI(x)
#' # End (Not run)------------
#' 
#' @seealso \code{\link[stats]{IQR}}, \code{\link[stats]{mad}}, \code{\link[DescTools]{MeanCI}}, \code{\link[DescTools]{MedianCI}}, \code{\link[DescTools]{Skew}}, \code{\link[DescTools]{Kurt}}
#' @note The basic functions in 'data_summary()' are cited from 'stats' package of R. \n Find more functions in 'DescTools' pacakge, such as Gini Coefficient, Geometric Mean and Standard Deviation, G-Test for Count Data, Skewness and Kurtosis, Confidence Interval for the Mean, Confidence Interval for the Median.
#' @references citation("stats")

data_summary <- function(x, digits = 3){
options(digits = digits)
n <- length(x)
m <- mean(x)
v <- var(x)
s <- sd(x)
me <- median(x)
q3 <- quantile(x,probs = 0.75)
q1 <- quantile(x,probs = 0.25)
min <- min(x)
max <- max(x)
cv <- 100*s/m
css <- sum((x-m)^2)
uss <- sum(x^2)
R <- max(x)-min(x)
IQR <- IQR(x)
mad <- mad(x)
sm <- s/sqrt(n)

shapiro.test(x)->test
print(test)
if (test$p.value>=0.05){
  message("\nThe data is Normal distribution (p-value >= 0.05). \nHence, it is recommended to use Mean +/- sd.\n\n")
}
else{message("\nThe data is non-Normal distribution (p-value < 0.05). \nHence, it is recommended to use Median (Q1-Q3).\n\n")
}

result1<-data.frame(N=n, Mean=m, Var=v, sd=s, se=sm, CV=cv, CSS=css, USS=uss, row.names="[1]")
result2<-data.frame(Median=me, mad=mad, Q1=q1, Q3=q3, IQR=IQR, Min = min, Max = max,Range=R, 
                   row.names="[2]")
print(result1)
cat("\n")
print(result2)
invisible(list(N=n, Mean=m, sd=s, se=sm, Median=me, Q1=q1, Q3=q3, IQR=IQR, Min = min, Max = max,Range=R, mad=mad, Var=v, CV=cv, CSS=css, USS=uss))
}